### DISCUSSION TITLE
Using server class for emulating PLC, data not available

### DISCUSSION THREAD

**Discussion by colingoodman** [gijzelaerr/python-snap7#439] [General]
I am attempting to stand up a fake PLC server for development purposes. I am having some trouble likely due to my lack of understanding with PLC devices generally.

I can stand up a server locally and connect to it fine, but attempting to populate it with some data fails.

```
server = snap7.server.Server()
server.start(tcpport=102)

db_data = (c_uint8 * 100)()
server.register_area(0, 1, db_data)
```

```
client = snap7.client.Client()
client.connect('127.0.0.1',0,1)
assert client.get_connected()
client.read_area(snap7.types.Areas.DB, 0, 100, 2)
```

It appears that my client has connected fine, though I get `RuntimeError: b'CPU : Item not available'` when trying to read the data I set via `register_area()`. I do not wish to use the default data provided just by running `python -m snap7.server`, I would like to specify my own.

Thanks in advance.

**Comment by colingoodman** [comment]
I've realized that the [server class](https://python-snap7.readthedocs.io/en/latest/_modules/snap7/server.html#Server) has a `mainloop()` function that appears to do almost what I want, I just need to change the values specified in `_init_standard_values()` to efficiently populate with whatever data I want.

**Comment by colingoodman** [comment]
I used the following to register some data at DB block 100 on the fake PLC server:
```python3
db_data = ctypes.c_int8 * len(bytearray(1000)
db_data = db_data.from_buffer(some_bytearray)

server.register_area(
    area_code=srvAreaDB,
    index=100,
    userdata=db_data
)
```

However, after a successful connection via my client, the following command 
`client.read_area(snap7.types.Areas.DB, 100, 0, 1)`
yields
`RuntimeError: b'CPU : Item not available'`

Does anyone know why the first line fails to properly register anything?

**Reply by nikteliy** [reply]
Hi @colingoodman
This code works for me:
```python
import snap7
import ctypes, time

db_data = ctypes.c_int8 * len(bytearray(1000))
db_data = db_data.from_buffer(bytearray(b"\xff" * 1000))
server = snap7.server.Server()
server.register_area(area_code=snap7.types.srvAreaDB, index=100, userdata=db_data)
server.start(tcpport=1102)
while True:
    while True:
        event = server.pick_event()
        if event:
            print(server.event_text(event))
        else:
            break
    time.sleep(1)
```

Here is the result of running this code in console:
![image](https://user-images.githubusercontent.com/52915342/229716405-c4132735-cf34-4552-9b3a-3767bfafaf74.png)

I hope this helps you find the error

**Reply by colingoodman** [reply]
Thanks for your reply, your verifying that it should work helped me out. I actually found that the issue was quite silly. I was using an environment variable to set the index parameter and forgot to convert the value to an integer. It may be worth noting that feeding `index` a string appears to silently cause issues.

**Comment by Gourav211** [comment]
This works for me.

I have pasted my client code : 

```
import snap7
from snap7.util import get_byte, set_byte
import time

from enum import IntEnum

class S7Area(IntEnum):
    PE = 0x81 # Peripheral Inputs ( read-only )
    PA = 0x82 # Peripheral Outputs ( read-only )
    MK = 0x83 # Merker (Memory) Area [Temporary, non-retentive memory (cleared on PLC restart)]( read/write )
    DB = 0x84 # Data Block ( read/write )
    CT = 0x1C # Counter ( read/write )
    TM = 0x1D # Timer ( read/write )

S7AreaDB = S7Area.DB # Data Block area for client operations

def main():
    client = snap7.client.Client()

    # Connect to server (assuming localhost and rack 0, slot 1)
    client.connect('127.0.0.1', 0, 1)

    if not client.get_connected():
        print("Failed to connect to server.")
        return

    print("Connected to Snap7 server.")

    db_number = 0
    start_byte = 10
    size = 1

    # Write value 42 to DB0 at byte 10
    write_buffer = bytearray(size)
    set_byte(write_buffer, 0, 42)
    client.write_area(S7AreaDB, db_number, start_byte, write_buffer)
    print(f"Wrote value 42 to DB{db_number}, byte {start_byte}")

    # Read back the value
    read_buffer = client.read_area(S7AreaDB, db_number, start_byte, size)
    read_value = get_byte(read_buffer, 0)
    print(f"Read value from DB{db_number}, byte {start_byte}: {read_value}")

    client.disconnect()
    print("Disconnected.")

if __name__ == "__main__":
    main()

```
